### PR TITLE
boards: Fix broken defconfigs

### PR DESCRIPTION
**Pull request by lucasssvaz**
## Summary

Normalize the broken configs that are breaking the CI.

## Impact

Normal CI behavior.

## Testing

CI pass.

### CODE REVIEW

**Review by pkarashchenko — APPROVED** [review]

**Review by pkarashchenko — APPROVED** [review]

**Comment by pkarashchenko** [comment]
Thank you for submitting this

**Comment by pkarashchenko** [comment]
@xiaoxiang781216 I see
```
====================================================================================
Configuration/Tool: stm32f334-disco/powerled,CONFIG_ARM_TOOLCHAIN_CLANG
------------------------------------------------------------------------------------
  Cleaning...
  Configuring...
...
  Disabling CONFIG_ARM_TOOLCHAIN_BUILDROOT
  Enabling CONFIG_ARM_TOOLCHAIN_CLANG
  Building NuttX...
ld.lld: error: section '.data' will not fit in region 'flash': overflowed by 32 bytes
ld.lld: error: section '.data' will not fit in region 'flash': overflowed by 80 bytes
ld.lld: error: section '.data' will not fit in region 'flash': overflowed by 92 bytes
ld.lld: error: section '.data' will not fit in region 'flash': overflowed by 92 bytes
make[1]: *** [Makefile:187: nuttx] Error 1
make: *** [tools/Unix.mk:520: nuttx] Error 2
make: Target 'all' not remade because of errors.
/github/workspace/sources/nuttx/tools/testbuild.sh: line 338: /github/workspace/sources/nuttx/../nuttx/nuttx.manifest: No such file or directory
  Normalize stm32f334-disco/powerled
```
But this is not treated as an error. Can somebody who has experience with Clang take a look into this?

**Comment by pkarashchenko** [comment]
Seems it is an error indeed, but CI just does not stop immediately. I've tried to build the same config with `CONFIG_ARM_TOOLCHAIN_GNU_EABI=y` and it succeeds.

**Comment by lucasssvaz** [comment]
> Seems it is an error indeed, but CI just does not stop immediately. I've tried to build the same config with `CONFIG_ARM_TOOLCHAIN_GNU_EABI=y` and it succeeds.

Should I enable it in this patch ?

**Comment by pkarashchenko** [comment]
It should be enabled in another place. I can do that in 2h (away from my laptop now)

**Comment by acassis** [comment]
@xiaoxiang781216 is there some way to avoid it happening in the future? A PR that disrupt the CI defconfig should be catch before it is too late, like happened here.

**Comment by pkarashchenko** [comment]
@acassis usually that is possible if changes are done in a single repo. But if we have dependency between apps and kernel code the current CI can't catch it.
There is a way to start using Google repo tool to track both apps and Nuttx, but that will change the flow that we will have to make a "tuple" change request like `[1/1]` for a single repository change, but `[1/2]` and `[2/2]` for multi repository change that should be a part of a same CI batch. So basically that is somehow similar to using apps as a submodule, but overcoming submodule limitation and track branch, but not a specific commit hash.

**Comment by pkarashchenko** [comment]
This is more a question to CI infrastructure for multi-repository projects

File: tools/ci/testlist/arm-10.dat [hunk 1]
```diff
@@ -4,7 +4,7 @@
 
 /arm/stm32/stm32f103-minimum,CONFIG_ARM_TOOLCHAIN_GNU_EABI
 
-/arm/stm32/stm32f334-disco,CONFIG_ARM_TOOLCHAIN_CLANG
+/arm/stm32/stm32f334-disco,CONFIG_ARM_TOOLCHAIN_GNU_EABI
```
Comment (pkarashchenko) [inline tools/ci/testlist/arm-10.dat:7]: Let's wait for CI to complete. If all is fine then let's separate this particular change into a separate commit.

**Comment by pkarashchenko** [comment]
Then I do not see any other way than reverting https://github.com/apache/nuttx/pull/8885

**Comment by pkarashchenko** [comment]
@xiaoxiang781216 @patacongo @acassis @jerpelea what do you think?

**Comment by raiden00pl** [comment]
we can try to add `CONFIG_DEFAULT_SMALL=y` for these broken configurations

**Comment by xiaoxiang781216** [comment]
> @xiaoxiang781216 @patacongo @acassis @jerpelea what do you think?

Another possible method is -Os plus LTO which may compensate the code size increased by https://github.com/apache/nuttx/pull/8885.

**Comment by patacongo** [comment]
It is inevitable that some configurations limited resource MCUs will eventually fail:  Gradual OS growth is expected and that is normal provided that growth is less than Moore's law.   Of course we don't advocate uncontrolled growth.  But older limited resource MCUs will eventually fail and it will not be possible to support them forever. 

We used to recommend 32Kb of FLASH for NuttX and there were many products fielded with 32Kb FLASH.  None of those are viable now.

Most growth is due to changes that support better POSIX compliance and Linux interoperability.  #8885 was necessary after commit e6227e19433c4999d500437d0a8a5c05f476ceb1 which expanded the number of standard signals from 19 to 29 and renumbered them to match Linux.  #8885 then recovered the 10 lost real time signals lost by that change.

I think it is a very bad idea to hold the OS back to support these older, limited platforms.  That is not the future.  We should revisit our value system and commit to a common vision.  What will it be like in 5 years?  Do we cripple it for obsolete parts?  Or do we embrace natural growth, improvement, and extension?

**Comment by raiden00pl** [comment]
Another idea is to disable NSH for the configurations than don't fit on FLASH and set ENTRYPOINT to the application which the config demonstrate. These boards with a small amount of FLASH don't use NSH anyway in practical applications because it is too big

**Comment by patacongo** [comment]
> Another possible method is -Os plus LTO which may compensate the code size increased by #8885.

There are several things I would need to know to have an opinion.  The size increase of #8885 should not have been very large (but I don't know know the exact size increase).  I would need to know that.  And I would need how close these platforms were to exceeding there memory resources before the change.  Looking at the small sizes of excesses, I suppose they were very close.

If the sizes were already close, then we will certainly exceed the capacity of those configurations on those platforms very soon anyway.  I don't believe that any measure to "work around the problem" for now is a meaningful solution for anything.  We have already exceed the capacity of those platforms for real world work and we are being disingenuous to advocate these configurations by keeping them in the repository.

I would vote to change nothing and just remove those configurations from repository.  That is honest and a proper representation of the state of the OS.  It is not just a bandaid for today.  It is a proper admission of the way things are and a proper solution for users of the OS.

File: tools/ci/testlist/arm-10.dat (hunk 1 above)
Comment (pkarashchenko) [reply tools/ci/testlist/arm-10.dat:7]: If the issue is solved by optimization so let's revert this chnage

File: boards/arm/stm32/nucleo-f302r8/configs/qenco/defconfig [hunk 2]
```diff
@@ -17,24 +71,40 @@ CONFIG_ARCH_INTERRUPTSTACK=1024
 CONFIG_ARCH_IRQPRIO=y
 CONFIG_BOARD_LOOPSPERMSEC=8499
 CONFIG_BUILTIN=y
-CONFIG_DEBUG_FULLOPT=y
```
Comment (pkarashchenko) [inline boards/arm/stm32/nucleo-f302r8/configs/qenco/defconfig:20]: @xiaoxiang781216 please correct me if I'm wrong, but isn't full opt the same as Os?

File: boards/arm/stm32/nucleo-f302r8/configs/qenco/defconfig [hunk 3]
```diff
@@ -17,24 +71,40 @@ CONFIG_ARCH_INTERRUPTSTACK=1024
...
 CONFIG_EXAMPLES_QENCODER_HAVE_MAXPOS=y
 CONFIG_EXAMPLES_QENCODER_MAXPOS=8192
+CONFIG_FILE_STREAM=y
 CONFIG_INIT_ENTRYPOINT="nsh_main"
 CONFIG_INTELHEX_BINARY=y
+CONFIG_LIBC_FLOATINGPOINT=y
+CONFIG_LIBC_LONG_LONG=y
+CONFIG_LIBM_TOOLCHAIN=y
 CONFIG_MQ_MAXMSGSIZE=5
+CONFIG_NSH_ALIAS=y
```
Comment (pkarashchenko) [inline boards/arm/stm32/nucleo-f302r8/configs/qenco/defconfig:88]: I think we definitely can turn this off

File: boards/arm/stm32/nucleo-f302r8/configs/qenco/defconfig [hunk 4]
```diff
@@ -17,24 +71,40 @@ CONFIG_ARCH_INTERRUPTSTACK=1024
...
 CONFIG_NSH_ARCHINIT=y
+CONFIG_NSH_ARGCAT=y
 CONFIG_NSH_BUILTIN_APPS=y
-CONFIG_PREALLOC_TIMERS=4
+CONFIG_NSH_CLE=y
+CONFIG_NSH_CMDOPT_HEXDUMP=y
+CONFIG_NSH_FILEIOSIZE=1024
+CONFIG_NSH_LINELEN=80
+CONFIG_NSH_QUOTE=y
```
Comment (pkarashchenko) [inline boards/arm/stm32/nucleo-f302r8/configs/qenco/defconfig:96]: I think that this also can be set to `n`

Comment (lucasssvaz) [reply boards/arm/stm32/nucleo-f302r8/configs/qenco/defconfig:96]: I'll test disabling these configs but as I edited my commented above, I think the only way will be to disable the NSH.

Comment (lucasssvaz) [reply boards/arm/stm32/nucleo-f302r8/configs/qenco/defconfig:96]: Maybe it is enough. No error was thrown
```
lucassvaz@Lucas-Espressif:~/nuttxspace/nuttx$ make -j
Create version.h
LN: platform/board to /home/lucassvaz/nuttxspace/apps/platform/dummy
Register: qe
Register: nsh
Register: sh
LD: nuttx
CP: nuttx.hex
CP: nuttx.bin
lucassvaz@Lucas-Espressif:~/nuttxspace/nuttx$ size nuttx
   text    data     bss     dec     hex filename
  64176     332    3252   67760   108b0 nuttx
```

Comment (pkarashchenko) [reply boards/arm/stm32/nucleo-f302r8/configs/qenco/defconfig:96]: Yes. Disable NSH is a good option. As well as disabling everything that is not used by application on the tiny platforms. Like if apps do not use message queues then those should be disabled. The same with POSIX timers, etc. In general I think most of those will be removed by -ffunctio-sections, but need to analyze map files to see that truth.

**Comment by patacongo** [comment]
If anyone cares where the size increases were with #8885 

    New Functions/Data (function size)
     38 T nxsig_nandset
     34 T sigandset
     34 T sigorset
     34 T nxsig_xorset
     18 T sigisemptyset
    ---
    124
    
    Not new, but apparently inlined before
     36 T sigaddset
     18 T sigfillset
      8 T sigemptyset
    ---
     62
    
    Modified Functions/Data (size increase)
    .text
    188 T nxsig_timedwait
    138 T nxsig_procmask
     86 T nxsig_deliver
     40 T nxsig_unmask_pendingsignal
     34 T nxsig_tcbdispatch
     18 T nxsig_delset
     18 T nxsig_pendingset
     18 T nxtask_setup_scheduler
     16 T nxsig_ismember
     16 T nxsig_addset
     16 T nxsig_cleanup
     16 t nxsig_alloc_block
     10 T nxsig_nanosleep
      4 T posix_spawnattr_init
    ---
    618
    
    Total FLASH:
    ---
    804
    
    .bss
     16 b g_idletcb

**Comment by acassis** [comment]
> 

Note that nxsig_timedwait and nxsig_procmask only represent also 50% of increase, maybe these could be the functions to investigate.

**Comment by patacongo** [comment]
> 

The three biggest increased in function size, are inexplicable.  There  is a very small change to each function but the expansion of the funciton size is large.  Something else is going on Look:

- Added three functioin calls to nxsig_timedwait: https://github.com/apache/nuttx/commit/717bb04cb7c6f1efb179d40a464e43e7cd13c7d8#diff-c610905678f957e3a394e88cfca1d9c97fc98017d9f5d4103d2707d07883b463
- Added one function call to nxsig_procmask:  https://github.com/apache/nuttx/commit/717bb04cb7c6f1efb179d40a464e43e7cd13c7d8#diff-9688e7f9a0a4d148691d3ab31ac7ebaa2df66c89270533ff21fbb9f969082948

**Review by xiaoxiang781216 — APPROVED** [review]

File: boards/arm/stm32/nucleo-f302r8/configs/qenco/defconfig (hunk 2 above)
Comment (pkarashchenko) [reply boards/arm/stm32/nucleo-f302r8/configs/qenco/defconfig:20]: Let's get back from custom opt to full opt

File: tools/ci/testlist/arm-10.dat (hunk 1 above)
Comment (lucasssvaz) [reply tools/ci/testlist/arm-10.dat]: Done 👍🏼

File: boards/arm/stm32/nucleo-f302r8/configs/qenco/defconfig (hunk 2 above)
Comment (lucasssvaz) [reply boards/arm/stm32/nucleo-f302r8/configs/qenco/defconfig:20]: Done 👍🏼

**Review by pkarashchenko — APPROVED** [review]

**Review by jerpelea — APPROVED** [review]

File: tools/ci/testlist/arm-10.dat (hunk 1 above)
Comment (lucasssvaz) [reply tools/ci/testlist/arm-10.dat]: I'll see what config fails and apply the same optimization to it

**Comment by acassis** [comment]
> > Note that nxsig_timedwait and nxsig_procmask alone represent almost 50% of increase, maybe these could be the functions to investigate.
> 
> The three biggest increases in function size, is inexplicable. There is a very small change to each function but the expansion of the funciton size is large. Something else is going on. One would probably have to look at when the linker is doing in the disassembly. Here are the code changes:
> 
>     * Added three functioin calls to nxsig_timedwait: [717bb04#diff-c610905678f957e3a394e88cfca1d9c97fc98017d9f5d4103d2707d07883b463](https://github.com/apache/nuttx/commit/717bb04cb7c6f1efb179d40a464e43e7cd13c7d8#diff-c610905678f957e3a394e88cfca1d9c97fc98017d9f5d4103d2707d07883b463)
> 
>     * Added one function call to nxsig_procmask:  [717bb04#diff-9688e7f9a0a4d148691d3ab31ac7ebaa2df66c89270533ff21fbb9f969082948](https://github.com/apache/nuttx/commit/717bb04cb7c6f1efb179d40a464e43e7cd13c7d8#diff-9688e7f9a0a4d148691d3ab31ac7ebaa2df66c89270533ff21fbb9f969082948)

Also many options that had a CONFIG flag to enable/disable were removed to force the kernel to become more POSIX compatible. An alternative could be having a flag CONFIG_NON_POSIX_COMPLIANT to allow disable some of these features (like the increased number of signals), this way we could have NuttX running again on MCUs with 32KB Flash for example.
Maybe in this case a NuttShell banner saying "Not POSIX Compliant" could be printed to warn the user.

Just some ideas, but of course, if we could keep the functionalities and just optimize the code to reduce the flash size, it will be much better!

**Pull request merged**